### DISCUSSION TITLE
Initial release of BresserWeatherSensorLW

### DISCUSSION THREAD

**Discussion by matthias-bs** [jgromes/RadioLib#1061] [Show and tell]
It's my pleasure to announce the initial release of [BresserWeatherSensorLW](https://github.com/matthias-bs/BresserWeatherSensorLW).

Many thanks to all people involved in adding LoRaWAN support to RadioLib, especially @jgromes, @StevenCellist and @HeadBoffin ! The quality of your code - including the examples - and documentation is excellent!

The [Persistence support for RadioLib](https://github.com/radiolib-org/radiolib-persistence/tree/main) allowed me to port my code from arduino-lmic to RadioLib in a very short time.

Special thanks for the excellent [RadioLib LoRaWAN on TTN starter script](https://github.com/jgromes/RadioLib/blob/master/examples/LoRaWAN/LoRaWAN_Starter/notes.md) - this will help new users a lot!

I'm sure RadioLib will soon be the base for countless LoRaWAN projects. 

Kind Regards,
Matthias

**Comment by StevenCellist** [comment]
Epic, _epic_, **epic**. Very cool to see this come alive & get adopted! Additional bonus points for excellent "Important notes" and "Features" in your readme.

Can we pretty please dip into your apparent RP2040 skills for the persistence library? As you may have seen, we are looking for an RP2040 deepsleep sketch that can serve as a starting point just like you found the ESP32 version :) If you can extract the necessary bits and pieces from your .ino sketch and submit as a PR, we will be very grateful!
Even more bonus points to be gained there by sharing with us the approximate power consumption you can achieve on your RP2040.. as that's usually a steep bump in the road!

**Reply by matthias-bs** [reply]
Yes, an RP2040 example is on my ToDo list. The implementation I currently have in my project is overly complicated, because it mimics the same behavior as with the ESP32 - reset after wake-up - which requires a few tricks.
The RP2040 specific implementation will just use the Preferences library for the nonces and
```
// RP2040 RAM is preserved during sleep; we just have to ensure that it is not initialized at startup (after reset)
uint8_t LWsession[RADIOLIB_LORAWAN_SESSION_BUF_SIZE] __attribute__((section(".uninitialized_data")));
```
Almost all code will be placed in `loop()`.
There is some room for improving power consumption by lowering the core voltage and maybe disabling RAM banks - I'm not sure if and when I will do this. 

I'd also like to provide an ESP32 example with a few things I added, like separation between LoRaWAN network handling and application, configuration downlink/uplink etc.

**Reply by StevenCellist** [reply]
Awesome, great to hear and looking forward!
Does RP2040 also have Preferences? Interesting, didn't know.

> I'm not sure if and when I will do this

Definitely not intended that you would, just curious to hear what you got :)

> I'd also like to provide an ESP32 example with a few things I added

Whilst definitely useful to have, I'm not completely sure where we would keep it - the Persistence repo is just, well, for persistence sketches ;) so you can maybe host it somewhere yourself, unless @jgromes has other plans

**Reply by matthias-bs** [reply]
[vshymanskyy/Preferences](https://github.com/vshymanskyy/Preferences) provides Preferences for RP2040. The API is fully compatible to the ESP32 version.

**Reply by Velocet** [reply]
Sorry to high jack this thread but it seems we all suffer from the same problem:
A kind of config database for LoRa modules with a sane naming scheme so projects can benefit from it.

https://github.com/Velocet/iown-homecontrol/tree/main/lib/LoRa32

This is just something i put together but i would fork this little lib into its own project (maybe under the new RadioLib-Org?) and incorporate ideas and configs from @matthias-bs plus some vendor specific extensions (will be kept in separate files like lora32_heltec.h).

Benefits:
- Same PIN naming for different boards/modules
- Scripts to distill the bare minimum for projects like RadioLib
- Vendor/Architecture specific extensions
  - Bugs (eg.: set pinMode to INPUT on GPI only pins used for interrupts)
  - Deep Sleep
  - Display config
  - ???

What do you think?

**Reply by HeadBoffin** [reply]
> What do you think?

Create your own discussion

**Reply by jgromes** [reply]
@Velocet I think it's worthwhile - though offtopic for this discussion, so please open up a new one so that we can figure out how to best approach this. Thanks!

**Comment by evon800c** [comment]
Hi @matthias-bs ,

it's nice to read this thread and look forward to new developments.

We exchanged a few messages for another yours project, I don't know if you remember.
I have both the board [Heltec WiFi LoRa 32 V3](https://heltec.org/project/wifi-lora-32-v3) and a bresser available.
If you need testing for this combination, I would be happy to help you in some way to say thank you for your efforts.

I would also like to thank @jgromes for his work on the [RadioLib](https://github.com/jgromes/RadioLib) library, which with the 6.5 version put in crisis my project, but which has made a huge evolution.

many thanks, Davide

**Reply by matthias-bs** [reply]
@evon800c Thanks for your message and your offer to test a setup!
